### PR TITLE
[FLINK-25432] Introduces ResourceCleaner to be used in the Dispatcher

### PR DESCRIPTION
**Pull request by XComp**
Related to [FLIP-194](https://cwiki.apache.org/confluence/display/FLINK/FLIP-194%3A+Introduce+the+JobResultStore)

## What is the purpose of the change

We want to introduce a generic interface for cleaning up job-related data in the `Dispatcher`. This affects the `BlobServer`, `HighAvailabilityServices`, `JobManagerMetricsGroup` and `JobGraphStore`. The goal is to move the cleanup logic into `ResourceCleaner` and remove complexity from the `Dispatcher` implementation.

## Brief change log

* `FileSystemBlobStore` was adapted to comply to the `BlobStore` interface for deletions
* Introduced interfaces for local and global cleanup
* Refactor `BlobServer`, `HighAvailabilityServices`, `JobManagerMetricsGroup` and `JobGraphStore` to implement the new interfaces
* Introduce `JobManagerRunnerRegistry` to wrap the HashMap holding the `JobManagerRunner` instances in the `Dispatcher`. This enables us to let `JobManagerRunnerRegistry` implement the new interfaces as well and to have the `JobManagerRunner.close` method being called through the cleanup.
* Adds `ResourceCleaner` interface and related helper/test classes
* Additionally, some minor hotfixes were added.

## Verifying this change

* These changes are mostly covered by existing tests that were just refactored to match the new interfaces
* A new unit test `JobManagerRunnerRegistryTest` was added to cover the `JobManagerRunnerRegistry` implementation
* We had to adapt the `DispatcherFailoverITCase` for it to succeed in the current state. This test might need a more general refactoring in some follow-up PR.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by XComp** [comment]
I'm not 100% sure about separating the cleanup into two interfaces. The reason I went for that approach was to make it easier to test `DefaultResourceCleaner`

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (XComp) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:228]: FYI: This was the reason the test didn't fail in the previous [FLINK-25430 PR](https://github.com/apache/flink/pull/18189) (`EmbeddedJobResultStore`)

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:181]: I miss how this change improves testability as we're only introducing new private constructors

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:252]: I'm not really a big fan of these safeguards.

- It's hard to keep them in sync when new code is added
- We should be able to enforce this in compile time simply by checking annotations (eg. https://github.com/uber/NullAway)
- It adds noise

What was the reason for adding these checks?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java:40]: It would be good to also inject the `ComponentMainThreadExecutor` here and assert that cleanupAsync is executed in the main thread as some implementation rely on it (eg. JobManagerRunnerRegistry)

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java:620]: why is this test no longer relevant?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DispatcherResourceCleanerFactoryTest.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DispatcherResourceCleanerFactoryTest.java:244]: Why can't we reuse `TestingHighAvailabilityServices` here? Same applies to other classes in this test.

The reasoning behind this is very same as with using `mockito`, if you decide to change something in the interface you have to go trough all implementations one by one.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:198]: should we simply poll the job result from the second dispatcher here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:126]:
```suggestion
                                    final Error error = temporaryErrorRef.getAndSet(null);
                                    if (error != null) {
                                        throw error;
                                    }
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:118]: Why error instead of a exception?

**Review by dmvk — COMMENTED** [review]
Thanks for the PR @XComp, this already looks pretty good 👍 I've added few comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/JobManagerRunnerRegistry.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/JobManagerRunnerRegistry.java:38]: Thinking about it, you're right: The decision on what to clean should lay solely in the `Dispatcher`. Having all components implement both interfaces adds some implicit dependency between the `Dispatcher` having these two different cleanup types. Instead, the `Dispatcher` should do a local cleanup as well when doing a global cleanup. 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/JobManagerMetricGroup.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/JobManagerMetricGroup.java:39]: Agree as already stated in [the comment below](https://github.com/apache/flink/pull/18536#discussion_r796427461). I will make change that.

File: flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/AbstractHaServicesTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/AbstractHaServicesTest.java:188]: I would leave the `internalJobCleanup` method signature like it is. The implementing classes don't use `IOException`, either. Changing the method signature to use `IOException` would mean that we have to do the `IOException` handling in each of the implementations. Instead, we can implement a general `IOException` handling in `AbstractHaServices.globalCleanup()`

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobGraphWriter.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobGraphWriter.java:43]: I moved it into `DefaultJobGraphStore.localCleanup` 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:931]: I reverted it back together with updating the `*alCleanup` method signatures... 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:950]: Good catch. I missed that. I'm going to add the copy constructor to `JobManagerRunnerRegistry.getRunningJobIds()` and add it analogously to `JobManagerRunnerRegistry.getJobManagerRunners()` to make these methods more robust towards `ConcurrentModificationExceptions`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:118]: I guess, the initial intention was to have `Error` as the selection condition in [ExceptionUtils.findThrowable(throwable, Error.class)](https://github.com/apache/flink/blob/master/flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java#L134) further down in the the test code. I left it like that because it doesn't cause any harm...

File: flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/TestingHighAvailabilityServices.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/TestingHighAvailabilityServices.java:295]: No reason other than keeping the old implementation. But that can be refactored 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/JobManagerRunnerRegistry.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/JobManagerRunnerRegistry.java:38]: FYI: Same applies to `HighAvailabilityServices`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java:620]: JobGraph removal and high availability cleanup are not depending on each other anymore.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:181]: I will revert that.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DispatcherResourceCleanerFactoryTest.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DispatcherResourceCleanerFactoryTest.java:244]: fair point. I refactored it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:198]: That would result in a timeout as well, wouldn't it? I guess that would be possible as well. But I didn't touch this code. The test is going to be modified once more in the upstream PR, anyway.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java]: I made the components thread-safe 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/GloballyCleanableResource.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/GloballyCleanableResource.java:35]: I cleaned the `*allyCleanableResource` interfaces up, made the implementing classes thread-safe, added MainThreadExecutor checks and updated the JavaDoc accordingly.

**Comment by XComp** [comment]
Thanks @dmvk . I addressed all your changes and put them in individual commits labeled as `[runtime][review]`. The commits labeled without `[runtime]`are the once refactoring the `*CleanableResource` interfaces removing the synchonous method, adding the MainThreadExecutor and making the implementations thread-safe

**Comment by XComp** [comment]
Rebased the branch fixing the `master` conflict

**Comment by XComp** [comment]
I had to add some fix to the last Dispatcher fix commit. The failure on submission wasn't handled properly first. This is fixed now.

**Comment by XComp** [comment]
There's another test failure `ZooKeeperDefaultDispatcherRunnerTest.testResourceCleanupUnderLeadershipChange` that is caused by the sequential call of local and global cleanup in the `DefaultGraphStore`. The local cleanup removes the `JobID` from `DefaultJobGraph.addedJobGraphs`. Global cleanup checks this set before removing data from the `StateHandleStore`. This doesn't happen because of the previously happened local cleanup call. We have to investigate whether we could just remove the contains check from the global cleanup

**Comment by XComp** [comment]
All review comments and the test failures are addressed. The PR is ready to get another review...

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/RunningInMainThreadAssert.java
Comment (XComp) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/RunningInMainThreadAssert.java:26]: I thought it would be helpful to have a simple function interface in cases like the `DefaultResourceCleaner` where we're not utilizing the `ComponentMainThreadExecutor` but just want to check whether the code is running in the mainthread.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:886]: Good point 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: I wasn't sure whether it would break the Github PR view here. That's why I added the revert. I will squash the commits properly after you give your thumbs up on squashing

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:441]: I found a different approach to reorganize this code in [3fc754f](https://github.com/apache/flink/commit/3fc754fb2f261ffd05f331a982f893dbf9b1828d). But I like you're suggestion. I will refactor it accordingly.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:905]: true, I also remove it from `Dispatcher.globalCleanup(JobID)` where it's also not really necessary. I will leave it in for `Dispatcher.localCleanup(JobID)`, though, because it's a requirement for the `DefaultResourceCleaner.cleanupAsync(JobID)` and having that assert makes it more explicit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:970]: Why not? 🤔

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:530]: true, I also remove it from `Dispatcher.globalCleanup(JobID)` where it's also not really necessary. I will leave it in for `Dispatcher.localCleanup(JobID)`, though, because it's a requirement for the `DefaultResourceCleaner.cleanupAsync(JobID)` and having that assert makes it more explicit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:1139]: Same as for the `Dispatcher.localCleanup(JobID)`: I thought of having this explicit safeguard here being more helpful because it's otherwise only visible through the `DefaultResourceCleaner.cleanupAsync` JavaDoc that it should be called from the main thread.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java
Comment (dmvk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java:259]: why do we need this? I thought we should always call either local or global cleanup, not both. What am I missing?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:970]: hmm, I've missed the last branch where we wrap the exception in IOException anyway :/ Non of the other branches will match anyway, so this basically has the same semantics as:

```suggestion
                throw new IOException(e);
```

which is more explicit 🤔

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/DefaultResourceCleaner.java:34]: Interesting approach. It definitely makes the public interface easier. I'm still a bit puzzled about how the translation from the `TriFunction` `FunctionalInterface` into the `BiFunction` `FunctionalInterface` works. But I have to do some reading about it

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStore.java:259]: true - I lacked some bigger picture here. The initial wrong-doing was that I removed the `GloballyCleanableResource` interface from the `JobManagerRunnerRegistry`. That required me to call `localCleanup` before calling the `globalCleanup` in the Dispatcher (because we still want to run the cleanup on the `JobManagerRunnerRegistry` even for the `globalCleanup`). 

Essentially, what we want to do is to make `JobManagerRunnerRegistry` implement `GloballyCleanableResource` as well (even though it doesn't have global resources). That enables the `Dispatcher` to call `localCleanup` and `globalCleanup` separately. 🤦

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/RunningInMainThreadAssert.java
Comment (dmvk) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/RunningInMainThreadAssert.java:26]: Let's remove it, it seems bit off to have this as a separate interface (is only intended for the ComponentMainThreadExecutor).

Comment (XComp) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/RunningInMainThreadAssert.java:26]: ok, already done 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:970]: Fair enough... 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/TestingResourceCleanerFactory.java
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/TestingResourceCleanerFactory.java:49]: rm

**Review by dmvk — APPROVED** [review]
LGTM 👍 This looks really good, thanks for the effort. Good job 🚀 🌔

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/TestingResourceCleanerFactory.java
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/TestingResourceCleanerFactory.java:49]: I removed the classes for now since they were not used...

**Comment by XComp** [comment]
Thanks for the review, @dmvk. I squashed the commits into reasonable chunks and rebased to `master`. I will merge the PR after the final AzureCI run

**Pull request merged**
